### ISSUE TITLE
[Problem] Assembly : No way to explode the assembly

### ISSUE THREAD

**Issue by PaddleStroke** [FreeCAD/FreeCAD#12419]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

It would be good to have a way to explode the assembly.

My current idea is to have a button 'explode assembly' that would temporarily change the placement to explode the assembly. Then pressing the tool again (or cancelling the task) would put things back in place.
Thoughts?

### Full version info

```shell
0.22
```

### Subproject(s) affected?

Other (specify in description)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by pierreporte** [comment]
Exploding an assembly isn’t just scattering the parts across the 3D space. It must be controlled in order to keep parts or sub-assemblies aligned. The explosion often needs to be partial. There should be the ability to have more than one exploded view (for assembly instructions). There should be lines to show where parts go (they must render in TechDraw). These lines shouldn’t be straight only.

The simple scattered expoded view is still useful so there could be two features (or sub-features): one for simple and usually temporary scattered view, one for more serious exploded views.

Depending on the strategy, it would be understandable to start small (scattered view) and to wait later for the more complex features.

Examples:

![image](https://github.com/FreeCAD/FreeCAD/assets/2029853/a89600bb-6495-455e-88b9-ae694cb943ca)

![image](https://github.com/FreeCAD/FreeCAD/assets/2029853/ee4448a5-b4e0-4133-942e-be31e423f6cb)

**Comment by maxwxyz** [comment]
As an additional idea, what about temporarily also disable all joints in the exploded state and being able to move individual parts around by dragging. (if the part in the default exploded state hides something)
It must be also ensured when the auto explode is active, that no part is disappearing or intersecting with other parts (all parts should be visible).

As @pierreporte mentioned I'm also in favor of a simple auto explode and a more complex version. Maybe one can define the direction of the individual part and the distance / rotation how far this should move and then save this parameters for the exploded state. (e.g. create an assembly for the exploded state and save this settings in the assembly / joints)

**Comment by PaddleStroke** [comment]
@qewer33 are you inspired to make an exploded view icon? :)

**Comment by qewer33** [comment]
> @qewer33 are you inspired to make an exploded view icon? :)

I'll give it a shot! Any ideas?

**Comment by maxwxyz** [comment]
What about the Assembly icon split apart?
@obelisk79 maybe interesting for DWG

**Comment by qewer33** [comment]
A couple of proof of concepts:

![image](https://github.com/FreeCAD/FreeCAD/assets/69015181/adea7370-4973-4c95-99c3-56163c7ca31c)
![image](https://github.com/FreeCAD/FreeCAD/assets/69015181/e07fd106-5b94-43a0-967c-df229e3f63fe)
![image](https://github.com/FreeCAD/FreeCAD/assets/69015181/e4da26ac-4e92-4e68-88d0-95e33fe46c4f)



[Assembly_ExplodedView.zip](https://github.com/FreeCAD/FreeCAD/files/14438333/Assembly_ExplodedView.zip)

**Comment by obelisk79** [comment]
I like the yellow and blue cylinder concept.

**Comment by maxwxyz** [comment]
> I like the yellow and blue cylinder concept.

But this could also be mistaken at small size for align axis or distance. Should we move this discussion to Discord?

**Comment by pierreporte** [comment]
@qewer33 I prefer the last one. It uses the assembly container icons and explodes its 3 parts.

**Comment by obelisk79** [comment]
That last one must have been an addtion... I like that the best so far

**Comment by maxwxyz** [comment]
Same. Top lightning of the bottom yellow cube seems a bit too dark?
IMO you could remove the red lines as they are almost invisible at smaller sizes and depending on your theme.

**Comment by pierreporte** [comment]
The red lines help understanding that the icon is for exploded views. They remain visible until the middle size, which is the one used for the toolbals. In the menu, the text helps further so it’s not a big deal if they are invisible.

**Comment by PaddleStroke** [comment]
Let's go with the last one for now then

**Comment by PaddleStroke** [comment]
@wwmayer I have a python binding question please. For this 'explode assembly view' tool I am trying to reuse the dragger (rather than reimplement the same thing in assembly).
Basically in this tool when I select a part I'm making a dragger appear on the part.
![image](https://github.com/FreeCAD/FreeCAD/assets/19984177/6a0c9891-1b90-4d15-967f-5dcbc61a11a2)

My problem is that I'm doing this in python : 

```
class ViewProviderExplodedView:
...
    def attach(self, vobj):
         ...
        self.dragger = coin.SoType.fromName("SoFCCSysDragger").createInstance()
        ...
```
But then the dragger object seems to not have all its c++ functions. For example `setUpAutoScale`.

So I am not sure how to handle this, because it seems that SoFCCSysDragger does not have a python interface. I can see different ideas : 
- Making a c++ class in assembly to interface the SoFCCSysDragger and make binding to this class.
- Make python interface for SoFCCSysDragger, but I don't know if it is possible ?
Thanks

**Comment by pierreporte** [comment]
For an efficient workflow, the user must be able to align the dragger with part features. It was briefly discussed in #12451.

**Comment by PaddleStroke** [comment]
Ah in fact it works it needs to be with * : 
`    return Base::Interpreter().createSWIGPointerObj("pivy.coin", "SoDragger *", asmDragger, 0);`

**Comment by PaddleStroke** [comment]
Interesting, I can get the dragger object in python now, but trying to add the callback crashes freecad when the event triggers.
```
        self.asmDragger = self.assembly.ViewObject.getDragger()
        self.asmDragger.addFinishCallback(self.draggerMoved)
```
Interestingly I finally found that this problem was described 10 years ago : https://forum.freecad.org/viewtopic.php?style=10&t=5594
So there seems to be a workaround : 
```
view = FreeCADGui.ActiveDocument.ActiveView
view.addDraggerCallback(self.dragger, "addFinishCallback", self.move_dragger)
view.removeDraggerCallback(self.dragger, "addFinishCallback", self.move_dragger)
```

I will try that and feedback if it works.

**Comment by PaddleStroke** [comment]
So adding the callbacks works.
```
        view.addDraggerCallback(self.asmDragger, "addFinishCallback", self.draggerFinished)
        view.addDraggerCallback(self.asmDragger, "addMotionCallback", self.draggerMoved)
```

But removing the callbacks crashes
```
        view.removeDraggerCallback(self.asmDragger, "addFinishCallback", self.draggerFinished)
        view.removeDraggerCallback(self.asmDragger, "addMotionCallback", self.draggerMoved)
```
It appears that SoFCCSysDragger is already adding a "addFinishCallback". So I'm adding a second one. And it appears `view.removeDraggerCallback` crashes in this case.

**Comment by wwmayer** [comment]
I haven't tested it but very likely you have to do it this way:
```
cbFin = view.addDraggerCallback(self.asmDragger, "addFinishCallback", self.draggerFinished)
cbMov = view.addDraggerCallback(self.asmDragger, "addMotionCallback", self.draggerMoved)
```
```
view.removeDraggerCallback(self.asmDragger, "addFinishCallback", cbFin)
view.removeDraggerCallback(self.asmDragger, "addMotionCallback", cbMov)
```

**Comment by PaddleStroke** [comment]
Thanks indeed it does not crash anymore!

**Comment by FEA-eng** [comment]
This was closed by the large Assembly PR so there should be a tool for exploded views in Assembly now, right? I can't find it in weekly. Will it be added later?

**Comment by maxwxyz** [comment]
I will add a PR to add the commands to the menu/toolbar

**Comment by PaddleStroke** [comment]
The new tools are currently hidden by default. you need to set the bool parameter 'ExperimentalFeatures' in Mod/Assembly to true to make them appear.
We have done this so that the tools can be thoroughly tested by power users before they become shown by default.
So I would suggest that we let the situation like this for a few months to iron bugs out.

**Comment by FEA-eng** [comment]
> So I would suggest that we let the situation like this for a few months to iron bugs out.

Ok, thanks for the information. But then they won't be available in 1.0 (feature freeze planned for the next week, release likely a few weeks later)?
